File: modules/wem-api/src/main/java/com/enonic/wem/api/content/page/PageTemplate.java
package com.enonic.wem.api.content.page;


import java.util.ArrayList;
import java.util.List;

import com.enonic.wem.api.content.Content;
import com.enonic.wem.api.data.Property;
import com.enonic.wem.api.data.RootDataSet;
import com.enonic.wem.api.data.Value;
import com.enonic.wem.api.schema.content.ContentTypeName;
import com.enonic.wem.api.schema.content.ContentTypeNames;

public final class PageTemplate
    extends Content
{
    private final PageTemplateKey key;

    private final PageTemplateName pageTemplateName;

    private PageTemplate( final Builder builder )
    {
        super( builder );
        this.key = builder.key;
        this.pageTemplateName = PageTemplateName.from( super.getName() );
    }

    public PageTemplateKey getKey()
    {
        return key;
    }

    public PageTemplateName getName()
    {
        return this.pageTemplateName;
    }

    public RootDataSet getConfig()
    {
        return getPage().getConfig();
    }

    public PageDescriptorKey getController()
    {
        if ( this.getPage() == null )
        {
            return null;
        }
        return this.getPage().getController();
    }

    public ContentTypeNames getCanRender()
    {
        final List<Property> supports = this.getContentData().getPropertiesByName( "supports" );
        final ArrayList<ContentTypeName> list = new ArrayList<>( supports.size() );
        for ( final Property property : supports )
        {
            if ( !property.hasNullValue() )
            {
                list.add( ContentTypeName.from( property.getString() ) );
            }
        }
        return ContentTypeNames.from( list );
    }

    public boolean canRender( ContentTypeName name )
    {
        return this.getCanRender().contains( name );
    }

    public boolean hasRegions()
    {
        return getPage().hasRegions();
    }

    public PageRegions getRegions()
    {
        return getPage().getRegions();
    }

    public static PageTemplate.Builder newPageTemplate()
    {
        return new Builder();
    }

    public static PageTemplate.Builder copyOf( final PageTemplate pageTemplate )
    {
        return new Builder( pageTemplate );
    }

    public static class Builder
        extends Content.Builder<Builder, PageTemplate>
    {
        private PageTemplateKey key;

        private Builder()
        {
            super();
        }

        private Builder( final PageTemplate source )
        {
            super( source );
        }

        public Builder key( final PageTemplateKey value )
        {
            this.key = value;
            this.contentId = value;
            return this;
        }

        public Builder controller( final PageDescriptorKey descriptorKey )
        {
            if ( this.page == null )
            {
                this.page = Page.newPage().
                    controller( descriptorKey ).
                    build();
            }
            else
            {
                this.page = Page.newPage( this.page ).
                    controller( descriptorKey ).
                    build();
            }
            return this;
        }

        public Builder canRender( final ContentTypeNames names )
        {
            for ( ContentTypeName name : names )
            {
                this.contentData.addProperty( "supports", Value.newString( name.toString() ) );
            }
            return this;
        }

        public Builder regions( final PageRegions value )
        {
            if ( this.page == null )
            {
                this.page = Page.newPage().
                    regions( value ).
                    build();
            }
            else
            {
                this.page = Page.newPage( this.page ).
                    regions( value ).
                    build();
            }

            return this;
        }

        public Builder config( final RootDataSet config )
        {
            if ( this.page == null )
            {
                this.page = Page.newPage().
                    config( config ).
                    build();
            }
            else
            {
                this.page = Page.newPage( this.page ).
                    config( config ).
                    build();
            }

            return this;
        }

        public PageTemplate build()
        {
            return new PageTemplate( this );
        }

    }

}

File: modules/wem-admin-ui/src/main/resources/web/admin/common/js/content/page/PageTemplate.ts
module api.content.page {

    import ContentTypeName = api.schema.content.ContentTypeName;
    import Property = api.data2.Property;
    import PropertyTree = api.data2.PropertyTree;
    import PropertyIdProvider = api.data2.PropertyIdProvider;

    export class PageTemplate extends api.content.Content implements api.Equitable {

        private canRender: ContentTypeName[];

        constructor(builder: PageTemplateBuilder) {

            super(builder);

            this.canRender = [];
            this.getContentData().forEachProperty("supports", (property: Property) => {
                this.canRender.push(new ContentTypeName(property.getString()));
            });
        }

        getKey(): PageTemplateKey {

            return <PageTemplateKey>this.getContentId();
        }

        getController(): DescriptorKey {

            return this.getPage().getController();
        }

        isCanRender(pattern: ContentTypeName): boolean {
            return this.getCanRender().some((name: ContentTypeName) => {
                return name.equals(pattern);
            });
        }

        getCanRender(): ContentTypeName[] {

            return this.canRender;
        }

        hasRegions(): boolean {
            return this.getPage().hasRegions();
        }

        getRegions(): PageRegions {
            return this.getPage().getRegions();
        }

        getConfig(): PropertyTree {
            return this.getPage().getConfig();
        }

        equals(o: api.Equitable): boolean {

            if (!api.ObjectHelper.iFrameSafeInstanceOf(o, PageTemplate)) {
                return false;
            }

            return super.equals(o);
        }

        clone(): PageTemplate {

            return this.newBuilder().build();
        }

        newBuilder(): PageTemplateBuilder {
            return new PageTemplateBuilder(this);
        }
    }

    export class PageTemplateBuilder extends api.content.ContentBuilder {

        constructor(source?: PageTemplate) {
            super(source);
        }

        fromContentJson(contentJson: api.content.json.ContentJson, propertyIdProvider: PropertyIdProvider): PageTemplateBuilder {
            super.fromContentJson(contentJson, propertyIdProvider);
            return this;
        }


        public build(): PageTemplate {
            return new PageTemplate(this);
        }
    }
}